### PR TITLE
Add criteria for minor and patch releases

### LINKED ISSUE #3528: 3.x.y patch release approach
We plan to release at least 3.0.4 and 3.1.1, and likely more 3.1.x releases.  We need to figure out what goes in and when to call these releases "done."

## Goal and Scope

The overarching goal of all 3.x(.y) release should be ***smoothing the path to Moonwalk.***  This has two components:

* Clarify ambiguities and omissions to make migration more predictable
* Reduce the guesswork involved in reaching full compliance with the spec for various use cases
 
Doing these things should increase satisfaction with the 3.x line as well.

We have a solid precedent for clarifying omissions or ambiguities from 3.0.3, which further specified `nullable`'s behavior to smooth the path to 3.1's full JSON Schema support:

* #2050
* #2046

This change technically invalidated assumptions that some implementations made, but did not invalidate any clear normative requirement.

We should also use normative language (MUST, SHOULD, MAY, etc.) wherever possible to make compliance requirements clear.  These requirements will need to map to Moonwalk, and ideally all compliance requirements could be tied to such language.  That's probably not realistic, but we can get closer.

## Definition of done

The main question here is whether 3.0.4 is the last 3.0.y or whether we can continue patching it.  If 3.0.4 is the last, then we want to get as many clarifications into it as possible (and 3.1.1 should go out with it, I think).

If we're OK doing more 3.0.y releases alongside 3.1.y releases, then we can pretty much ship a patch release whenever we think there's something worth shipping.  A more-or-less quarterly cadence is probably ~~realistic~~ _[EDIT: that's probably more frequent than necessary]_ (at least once we solve our release process roadblocks).

I personally don't see the need to end the 3.0.y line.  It and the 3.1.y line are close enough that most patches will apply to both.  It would be a minimal effort to make 3.0 customers feel supported.

### LINKED ISSUE #3529: 3.x minor release approach
We know we want to do a 3.2 release, and talked in this week's TDC Call about the possibility of further 3.x releases.

As with the patch releases (see #3528), I think the overarching goal should be ***smoothing the path to Moonwalk***.  I mean this both technically and socially, regarding both the user and implementer communities.

## Rebuild trust

_All of this is very much my personal opinion, and not backed up by formal market research or anything similar._

OpenAPI has never truly produced a minor release.  2.0 came out in 2014, 3.0 in 2017, and 3.1 in 2021.  4.0 is planned for the end of 2024.  All required massive amounts of work to support, which is part of why the releases are spaced 3-4 years apart.  That is how long it has taken the ecosystem to catch up enough to even consider another major version.

While some in our community are really excited about Moonwalk, others are lamenting yet another big change that probably won't benefit them for quite a few years (although I am hoping we can do a variety of things to make Moonwalk easier to support and adopt).

The experience of the OpenAPI community has been one of multiple years of no visible movement, punctuated by massive changes that aren't broadly supported until the next massive change appears.  Reading through the issue backlog, there's a lot of frustration at both the lack of interim progress and the fact that long adoption timelines mean that closing an issue as "resolved" doesn't mean it will be usable anytime soon.

### 3.2:  the critical starting point

The most important thing about 3.2 is that it should be _very easy to implement if you already support 3.1_.  It should deliver at least one exciting feature, and probably no more than three.  It shouldn't break anything (aside from assumptions regarding ambiguous requirements, [as discussed for patch releases](https://github.com/OAI/OpenAPI-Specification/issues/3528).

* We want a release that vendors are happy to immediately support, that benefits a significant chunk of users.
* We want people who are investing, or considering investing in 3.1, to see that they will see additional benefits from the 3.x line while vendors catch up to Moonwalk.

With 3.2, smaller is better!

### 3.3+: converging with Moonwalk

With a successful 3.2, we can talk about a similarly-sized 3.3 without kicking up too much fear of another 3.1.  By this time, Moonwalk should be increasingly clear  We should consider "backporting" features that can fit into the 3.x syntax, much like Python 2.x minor releases tended to converge with 3.x releases for quite some time.

Obviously, a lot of Moonwalk can't be backported – the same was true of Python 3.  But I suspect we will find some things that can.  In terms of the basic semantics if not all of the details.

As with 3.2, any 3.x release should add at least one and probably no more than three features, to ensure that they can be implemented and used before the next 3.x+1 comes out.

## Release cadence

Assuming we do multiple 3.x releases, the cadence should allow implementers who actively start working on a new 3.x release time to deploy the new support before piling on another minor release.  Since these release should (as proper minor releases) be purely additive, it wouldn't be bad if they crowd up a bit- implementing 3.x+1 ought to mean you automatically can support 3.x (for x > 0).

### PR DESCRIPTION
**Pull request by lornajane**
We had some great discussions in the TDC meetings when we were finalising what should go into the recent patch releases and what must be a minor release. This pull request takes those discussions into our CONTRIBUTING file.

Fixes #3528 
Fixes #3529

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -153,10 +153,34 @@ The following additional rules should be followed but currently are not enforced
...
+Changes in minor releases (such as 3.2, 3.3) meet the following criteria:
+
+* Are **backwards-compatible** and be reasonably easy to implement in tooling that already supports the previous minor version.
+  For example, new optional fields can be added.
+* Drive quality-of-life improvements to support how OpenAPI is used by practitioners, so that OpenAPI evolves to continue to meet user needs.
+  For example, adding fields to support changes in other standards, or adopting common `x-*` extension fields into the specification.
+* Bring the future closer by making changes that are in line with future 3.x releases and the planned OpenAPI 4.x (Moonwalk) specification as the details of that become available.
+* Make the specification document clearer or easier to understand.
+
+A minor release is due when there are some meaningful features (including one or a small nummber of headline features).
```
Comment (ralfhandl) [inline CONTRIBUTING.md:171]:
```suggestion
A minor release is due when there are some meaningful features (including one or a small number of headline features).
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -153,10 +153,34 @@ The following additional rules should be followed but currently are not enforced
...
+
+### Patch Releases
+
+Patch releases reflect a constant quest for improving the active minor versions of OpenAPI.
+Since we do not edit specification documents after publication, even the smallest change has to be in a new release.
+
+Changes in patch releases meet the following criteria:
+
+* Minor updates such as spelling or formatting fixes, including link updates.
```
Comment (ralfhandl) [inline CONTRIBUTING.md:180]:
```suggestion
* Editorial changes such as spelling or formatting fixes, including link updates.
```

**Review by ralfhandl — COMMENTED** [review]
+1, minor nits

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -33,7 +33,7 @@ As of October 2024 (post-OAS 3.0.4 and 3.1.1), the OAS is developed in the `src/
...
 For example, if a change in November 2024 apples to both 3.1 and 3.2, the PR would go to the `v3.1-dev` branch, which will be merged up to `v3.2-dev` before the next 3.2 release.
 
-Releases are published to the [spec site](https://spec.opanapis.org) by creating an `X.Y.Z-rel` branch where `src/oas.md` is renamed to the appropriate `versions/X.Y.Z.md` file and them merged to `main`.  The HTML versions of the OAS are automatically generated from the `versions` directory on `main`.  This renaming on the `X.Y.Z-rel` branch preserves the commit history for the published file on `main` when using `git log --follow` (as is the case for all older published files).
+Releases are published to the [spec site](https://spec.openapis.org) by creating an `X.Y.Z-rel` branch where `src/oas.md` is renamed to the appropriate `versions/X.Y.Z.md` file and them merged to `main`.  The HTML versions of the OAS are automatically generated from the `versions` directory on `main`.  This renaming on the `X.Y.Z-rel` branch preserves the commit history for the published file on `main` when using `git log --follow` (as is the case for all older published files).
```
Comment (ralfhandl) [inline CONTRIBUTING.md:36]:
```suggestion
Releases are published to the [spec site](https://spec.openapis.org) by creating an `X.Y.Z-rel` branch where `src/oas.md` is renamed to the appropriate `versions/X.Y.Z.md` file and then merged to `main`.  The HTML versions of the OAS are automatically generated from the `versions` directory on `main`.  This renaming on the `X.Y.Z-rel` branch preserves the commit history for the published file on `main` when using `git log --follow` (as is the case for all older published files).
```
Typo spotted by @jeremyfiel in #4186

**Review by mikekistler — DISMISSED** [review]
Looks good. 👍

File: CONTRIBUTING.md (hunk 3 above)
Comment (lornajane) [reply CONTRIBUTING.md:36]: If this has been added elsewhere then it should be accepted in that pull request. If we make the same change here, we may cause conflicts?

Comment (ralfhandl) [reply CONTRIBUTING.md:36]: I would have guessed no, and guessed that we'll get a conflict if we don't make the same change here.

The second guess has been confirmed, the first remains unproven 😎

**Review by ralfhandl — APPROVED** [review]

**Pull request merged**